src/compiler/checker.ts

```diff
@@ -27291,14 +27290,21 @@ namespace ts {
         // determining the type of foo would cause foo to be given type any because of the recursive reference.
         // Delaying the type check of the body ensures foo has been assigned a type.
         function checkNodeDeferred(node: Node) {
-            if (deferredNodes) {
+            const enclosingFile = getSourceFileOfNode(node);
+            const links = getNodeLinks(enclosingFile);
+            if (!(links.flags & NodeCheckFlags.TypeChecked)) {
+                links.deferredNodes = links.deferredNodes || createMap();
                 const id = "" + getNodeId(node);
-                deferredNodes.set(id, node);
+                links.deferredNodes.set(id, node);
             }

```

Comment:
Should we assert in the else case that deferred node already exists?

Reply:
No. We are perfectly fine adding the node multiple times (and in fact we will in many cases), we're still only visiting it once for errors in the end. That's why we swapped to a map in the first place, instead of an array.

Reply:
I meant when file was typechecked? The assert would be to verify the deferredNode exists in the map.

Reply:
Seeings as previously if `deferredNodes` didn't exist, we'd quietly do nothing, I don't think an assert is quite correct. And experimentally, it looks like such an assert is only triggered during a full typecheck walk after a normal check (like in tests) under scenarios with erroneous syntax, such as a top-level return statement.